File: tests/benchmark_stratified.py
"""
Benchmark: C++ vs Python Stratified Cox Implementation

Compares performance across different scenarios:
- Number of strata
- Number of observations
- With/without ties
- With/without start times
"""

import numpy as np
import time
from coxdev import StratifiedCoxDeviance, StratifiedCoxDevianceCpp


def benchmark_single(n, n_strata, n_calls=100, with_ties=False, with_start=False, seed=42):
    """Run a single benchmark comparison."""
    np.random.seed(seed)

    if with_ties:
        # Create data with many ties
        event = np.random.choice(np.arange(1, n // 10 + 1), n).astype(float)
    else:
        event = np.random.exponential(5, n)

    status = np.random.binomial(1, 0.7, n)
    strata = np.random.choice(n_strata, n)
    eta = np.random.randn(n) * 0.5
    weights = np.random.uniform(0.5, 2.0, n)

    if with_start:
        start = np.random.uniform(0, 2, n)
        event = start + np.abs(event)  # Ensure event > start
    else:
        start = None

    # Create objects (preprocessing time)
    t0 = time.perf_counter()
    py_cox = StratifiedCoxDeviance(event=event, status=status, strata=strata, start=start)
    py_init_time = time.perf_counter() - t0

    t0 = time.perf_counter()
    cpp_cox = StratifiedCoxDevianceCpp(event=event, status=status, strata=strata, start=start)
    cpp_init_time = time.perf_counter() - t0

    # Warm up
    py_cox(eta, weights)
    cpp_cox(eta, weights)

    # Benchmark deviance computation
    t0 = time.perf_counter()
    for _ in range(n_calls):
        py_result = py_cox(eta, weights)
    py_call_time = (time.perf_counter() - t0) / n_calls

    t0 = time.perf_counter()
    for _ in range(n_calls):
        cpp_result = cpp_cox(eta, weights)
    cpp_call_time = (time.perf_counter() - t0) / n_calls

    # Benchmark information matvec
    v = np.random.randn(n)
    py_info = py_cox.information(eta, weights)
    cpp_info = cpp_cox.information(eta, weights)

    t0 = time.perf_counter()
    for _ in range(n_calls):
        _ = py_info @ v
    py_matvec_time = (time.perf_counter() - t0) / n_calls

    t0 = time.perf_counter()
    for _ in range(n_calls):
        _ = cpp_info @ v
    cpp_matvec_time = (time.perf_counter() - t0) / n_calls

    # Verify results match
    assert np.isclose(py_result.deviance, cpp_result.deviance, rtol=1e-10)

    return {
        'n': n,
        'n_strata': n_strata,
        'with_ties': with_ties,
        'with_start': with_start,
        'py_init_ms': py_init_time * 1000,
        'cpp_init_ms': cpp_init_time * 1000,
        'py_call_ms': py_call_time * 1000,
        'cpp_call_ms': cpp_call_time * 1000,
        'py_matvec_ms': py_matvec_time * 1000,
        'cpp_matvec_ms': cpp_matvec_time * 1000,
        'init_speedup': py_init_time / cpp_init_time if cpp_init_time > 0 else float('inf'),
        'call_speedup': py_call_time / cpp_call_time if cpp_call_time > 0 else float('inf'),
        'matvec_speedup': py_matvec_time / cpp_matvec_time if cpp_matvec_time > 0 else float('inf'),
    }


def run_benchmarks():
    """Run comprehensive benchmarks."""
    print("=" * 80)
    print("BENCHMARK: C++ vs Python Stratified Cox Implementation")
    print("=" * 80)

    results = []

    # Varying number of observations with fixed strata
    print("\n### Varying n (observations), fixed 10 strata ###")
    print(f"{'n':>8} | {'Python (ms)':>12} | {'C++ (ms)':>12} | {'Speedup':>8}")
    print("-" * 50)
    for n in [100, 500, 1000, 5000, 10000]:
        r = benchmark_single(n, n_strata=10, n_calls=50)
        results.append(r)
        print(f"{n:>8} | {r['py_call_ms']:>12.4f} | {r['cpp_call_ms']:>12.4f} | {r['call_speedup']:>8.2f}x")

    # Varying number of strata with fixed observations
    print("\n### Varying strata count, fixed n=1000 ###")
    print(f"{'strata':>8} | {'Python (ms)':>12} | {'C++ (ms)':>12} | {'Speedup':>8}")
    print("-" * 50)
    for n_strata in [2, 5, 10, 50, 100, 200]:
        r = benchmark_single(n=1000, n_strata=n_strata, n_calls=50)
        results.append(r)
        print(f"{n_strata:>8} | {r['py_call_ms']:>12.4f} | {r['cpp_call_ms']:>12.4f} | {r['call_speedup']:>8.2f}x")

    # Many small strata (worst case for Python loop overhead)
    print("\n### Many small strata (5 obs each) ###")
    print(f"{'strata':>8} | {'Python (ms)':>12} | {'C++ (ms)':>12} | {'Speedup':>8}")
    print("-" * 50)
    for n_strata in [10, 50, 100, 200, 500]:
        n = n_strata * 5
        r = benchmark_single(n=n, n_strata=n_strata, n_calls=50)
        results.append(r)
        print(f"{n_strata:>8} | {r['py_call_ms']:>12.4f} | {r['cpp_call_ms']:>12.4f} | {r['call_speedup']:>8.2f}x")

    # With ties (Efron)
    print("\n### With many ties (Efron), n=1000, 10 strata ###")
    r = benchmark_single(n=1000, n_strata=10, n_calls=50, with_ties=True)
    results.append(r)
    print(f"Python: {r['py_call_ms']:.4f} ms | C++: {r['cpp_call_ms']:.4f} ms | Speedup: {r['call_speedup']:.2f}x")

    # With start times (left truncation)
    print("\n### With start times (left truncation), n=1000, 10 strata ###")
    r = benchmark_single(n=1000, n_strata=10, n_calls=50, with_start=True)
    results.append(r)
    print(f"Python: {r['py_call_ms']:.4f} ms | C++: {r['cpp_call_ms']:.4f} ms | Speedup: {r['call_speedup']:.2f}x")

    # Information matrix matvec
    print("\n### Information matrix matvec, n=1000, 10 strata ###")
    r = benchmark_single(n=1000, n_strata=10, n_calls=50)
    print(f"Python: {r['py_matvec_ms']:.4f} ms | C++: {r['cpp_matvec_ms']:.4f} ms | Speedup: {r['matvec_speedup']:.2f}x")

    # Initialization time
    print("\n### Initialization (preprocessing) time, n=5000, 50 strata ###")
    r = benchmark_single(n=5000, n_strata=50, n_calls=10)
    print(f"Python: {r['py_init_ms']:.4f} ms | C++: {r['cpp_init_ms']:.4f} ms | Speedup: {r['init_speedup']:.2f}x")

    print("\n" + "=" * 80)
    print("SUMMARY")
    print("=" * 80)

    # Compute average speedups
    call_speedups = [r['call_speedup'] for r in results]
    print(f"Call speedup range: {min(call_speedups):.2f}x - {max(call_speedups):.2f}x")
    print(f"Mean call speedup: {np.mean(call_speedups):.2f}x")

    return results


if __name__ == "__main__":
    run_benchmarks()

File: R_pkg/coxdev/inst/benchmark_stratified.R
#!/usr/bin/env Rscript
#' Benchmark: C++ Stratified vs R Loop Implementation
#'
#' Compares the C++ stratified Cox implementation against
#' an R loop that calls the unstratified version per stratum
#' (similar to what glmnet does internally).

library(coxdev)

# R loop implementation (baseline - similar to glmnet's approach)
stratified_r_loop <- function(event, status, strata, eta, weights = NULL,
                               start = NULL, tie_breaking = "efron") {
  unique_strata <- sort(unique(strata))
  n <- length(event)
  if (is.null(weights)) weights <- rep(1, n)

  total_dev <- 0
  gradient <- numeric(n)
  diag_hessian <- numeric(n)


  for (s in unique_strata) {
    idx <- which(strata == s)
    cox_s <- make_cox_deviance(
      event = event[idx],
      status = status[idx],
      start = if (!is.null(start)) start[idx] else NULL,
      tie_breaking = tie_breaking
    )
    result <- cox_s$coxdev(eta[idx], weights[idx])
    total_dev <- total_dev + result$deviance
    gradient[idx] <- result$gradient
    diag_hessian[idx] <- result$diag_hessian
  }

  list(deviance = total_dev, gradient = gradient, diag_hessian = diag_hessian)
}

benchmark_single <- function(n, n_strata, n_calls = 50, with_ties = FALSE,
                              with_start = FALSE, seed = 42) {
  set.seed(seed)

  if (with_ties) {
    event <- sample(1:(n/10), n, replace = TRUE)
  } else {
    event <- rexp(n, rate = 0.2)
  }

  status <- rbinom(n, 1, 0.7)
  strata <- sample(1:n_strata, n, replace = TRUE)
  eta <- rnorm(n) * 0.5
  weights <- runif(n, 0.5, 2.0)

  if (with_start) {
    start <- runif(n, 0, 2)
    event <- start + abs(event)
  } else {
    start <- NULL
  }

  # Initialize C++ version
  cpp_init_time <- system.time({
    cpp_cox <- make_stratified_cox_deviance(
      event = event,
      status = status,
      strata = strata,
      start = start
    )
  })["elapsed"]

  # Warm up
  cpp_cox$coxdev(eta, weights)
  stratified_r_loop(event, status, strata, eta, weights, start)

  # Benchmark C++ calls
  cpp_call_time <- system.time({
    for (i in 1:n_calls) {
      cpp_result <- cpp_cox$coxdev(eta, weights)
    }
  })["elapsed"] / n_calls

  # Benchmark R loop calls
  r_call_time <- system.time({
    for (i in 1:n_calls) {
      r_result <- stratified_r_loop(event, status, strata, eta, weights, start)
    }
  })["elapsed"] / n_calls

  # Verify results match
  stopifnot(abs(cpp_result$deviance - r_result$deviance) < 1e-10)

  list(
    n = n,
    n_strata = n_strata,
    with_ties = with_ties,
    with_start = with_start,
    cpp_init_ms = cpp_init_time * 1000,
    cpp_call_ms = cpp_call_time * 1000,
    r_call_ms = r_call_time * 1000,
    speedup = r_call_time / cpp_call_time
  )
}

run_benchmarks <- function() {
  cat(strrep("=", 80), "\n")
  cat("BENCHMARK: C++ Stratified vs R Loop Implementation\n")
  cat(strrep("=", 80), "\n\n")

  results <- list()

  # Varying number of observations with fixed strata
  cat("### Varying n (observations), fixed 10 strata ###\n")
  cat(sprintf("%8s | %12s | %12s | %8s\n", "n", "R loop (ms)", "C++ (ms)", "Speedup"))
  cat(strrep("-", 50), "\n")
  for (n in c(100, 500, 1000, 5000, 10000)) {
    r <- benchmark_single(n, n_strata = 10, n_calls = 30)
    results <- c(results, list(r))
    cat(sprintf("%8d | %12.4f | %12.4f | %7.2fx\n",
                n, r$r_call_ms, r$cpp_call_ms, r$speedup))
  }

  # Varying number of strata with fixed observations
  cat("\n### Varying strata count, fixed n=1000 ###\n")
  cat(sprintf("%8s | %12s | %12s | %8s\n", "strata", "R loop (ms)", "C++ (ms)", "Speedup"))
  cat(strrep("-", 50), "\n")
  for (n_strata in c(2, 5, 10, 50, 100, 200)) {
    r <- benchmark_single(n = 1000, n_strata = n_strata, n_calls = 30)
    results <- c(results, list(r))
    cat(sprintf("%8d | %12.4f | %12.4f | %7.2fx\n",
                n_strata, r$r_call_ms, r$cpp_call_ms, r$speedup))
  }

  # Many small strata (worst case for R loop overhead)
  cat("\n### Many small strata (5 obs each) ###\n")
  cat(sprintf("%8s | %12s | %12s | %8s\n", "strata", "R loop (ms)", "C++ (ms)", "Speedup"))
  cat(strrep("-", 50), "\n")
  for (n_strata in c(10, 50, 100, 200, 500)) {
    n <- n_strata * 5
    r <- benchmark_single(n = n, n_strata = n_strata, n_calls = 30)
    results <- c(results, list(r))
    cat(sprintf("%8d | %12.4f | %12.4f | %7.2fx\n",
                n_strata, r$r_call_ms, r$cpp_call_ms, r$speedup))
  }

  # With ties (Efron)
  cat("\n### With many ties (Efron), n=1000, 10 strata ###\n")
  r <- benchmark_single(n = 1000, n_strata = 10, n_calls = 30, with_ties = TRUE)
  results <- c(results, list(r))
  cat(sprintf("R loop: %.4f ms | C++: %.4f ms | Speedup: %.2fx\n",
              r$r_call_ms, r$cpp_call_ms, r$speedup))

  # With start times (left truncation)
  cat("\n### With start times (left truncation), n=1000, 10 strata ###\n")
  r <- benchmark_single(n = 1000, n_strata = 10, n_calls = 30, with_start = TRUE)
  results <- c(results, list(r))
  cat(sprintf("R loop: %.4f ms | C++: %.4f ms | Speedup: %.2fx\n",
              r$r_call_ms, r$cpp_call_ms, r$speedup))

  # Large scale test
  cat("\n### Large scale: n=50000, 500 strata ###\n")
  r <- benchmark_single(n = 50000, n_strata = 500, n_calls = 10)
  results <- c(results, list(r))
  cat(sprintf("R loop: %.4f ms | C++: %.4f ms | Speedup: %.2fx\n",
              r$r_call_ms, r$cpp_call_ms, r$speedup))

  cat("\n", strrep("=", 80), "\n")
  cat("SUMMARY\n")
  cat(strrep("=", 80), "\n")

  speedups <- sapply(results, function(r) r$speedup)
  cat(sprintf("Speedup range: %.2fx - %.2fx\n", min(speedups), max(speedups)))
  cat(sprintf("Mean speedup: %.2fx\n", mean(speedups)))

  invisible(results)
}

# Run if executed directly
if (!interactive()) {
  run_benchmarks()
}
